Update timemory to support new dependencies and expose more build options

jrmadsen opened the pull request.

jrmadsen committed:
Update timemory/package.py to support new dependencies and expose more build options

jrmadsen committed:
Added missing 'use_arch' option handling

jrmadsen committed:
Fixes for python 2.6

jrmadsen committed:
Fix pyflake error F632

jrmadsen committed:
F632 flake fixes

adamjstewart commented on var/spack/repos/builtin/packages/timemory/package.py:
```diff
@@ -17,23 +18,66 @@ class Timemory(CMakePackage):
     version('3.0.0', commit='b36b1673b2c6b7ff3126d8261bef0f8f176c7beb',
             submodules=True)
+    version('3.0.1', commit='a5bb58b5e4d44b71f699f536ad1b56722f213ce6',
+            submodules=True)
```
We want to keep versions in order of newest to oldest

jrmadsen committed:
Update package.py

jrmadsen committed:
Update package.py

adamjstewart commented on var/spack/repos/builtin/packages/timemory/package.py:
```diff
@@ -22,18 +23,61 @@ class Timemory(CMakePackage):
+    linux = False if platform == 'darwin' else True
+
+    variant('shared', default=True, description='Build shared libraries')
+    variant('static', default=False, description='Build static libraries')
```
You'll prob want a conflict for `~shared~static`, as I assume that combination isn't possible.

jrmadsen replied:
It is actually possible. It is a template library so it can actually function as header only although it's not really recommended bc w/o any extern template instantiations the compile times get pretty long when downstream applications use it.

adamjstewart commented on var/spack/repos/builtin/packages/timemory/package.py:
```diff
@@ -22,18 +23,61 @@ class Timemory(CMakePackage):
+            values=('auto', 'kepler', 'tesla', 'maxwell', 'pascal',
+                    'volta', 'turing'), multi=False)
+    variant('cpu_target', default='auto',
+            description='Build for specific cpu architecture (specify cpu-model)')
```
Spack already has microarchitecture detection. I would use that instead of a variant. @alalazo or @becker33 can explain how it works.

jrmadsen replied:
While that sounds interesting, I omitted listing the 20+ cpu variants handled here: https://github.com/NERSC/timemory/blob/develop/cmake/Modules/FindCpuArch.cmake
and the cuda values are specific to what is here: https://github.com/NERSC/timemory/blob/5c0f89880db25e31a0778d8252b05c4ad5e50803/cmake/Modules/CUDAConfig.cmake#L30.

adamjstewart commented on var/spack/repos/builtin/packages/timemory/package.py:
```diff
@@ -22,18 +23,61 @@ class Timemory(CMakePackage):
+            description='Build all of timemory w/ cpu_target architecture flags (default: roofline toolkit only)')
+    variant('tls_model', default='global-dynamic', description='Thread-local static model',
+            values=('global-dynamic', 'local-dynamic', 'initial-exec', 'local-exec'),
+            multi=False)
```
I would prefer that variants (and possible values) use underscores instead of dashes to prevent confusion. Boolean variants can be disabled with `-foo`, so variants with dashes in the name are misleading.

jrmadsen replied:
I understand that. The problem is that those values actually get translated to this directly from the CMake option to a C++ compiler flag: `add_cxx_flag_if_avail("-ftls-model=${TIMEMORY_TLS_MODEL}")` so it would necessitate doing something like `tls = tls.replace('_', '-')` which is the sort of generic string substitution I try to avoid whenever possible. I'll change it if you feel strongly but unless it has the potential to confuse the parser, I'd prefer to keep it as is.

adamjstewart replied:
Oh, it won't confuse the parser, it just might confuse humans. It's your package, so you can do things how you want, but I don't think that string substitution will break unless you need options with both `-` and `_` in them.

jrmadsen replied:
Yea I think anyone that actually wants to tweak this (advanced) option will likely be familiar with the options already

adamjstewart commented on var/spack/repos/builtin/packages/timemory/package.py:
```diff
@@ -48,77 +92,83 @@ class Timemory(CMakePackage):
-            '-DTIMEMORY_BUILD_GOTCHA=OFF',
-            '-DTIMEMORY_BUILD_CALIPER=OFF',
-            '-DTIMEMORY_BUILD_TOOLS=ON',
+            '-DTIMEMORY_BUILD_PYTHON=ON',
```
Shouldn't this remain conditional based on whether or not `+python` is requested? Python is only a dependency when `+python` is requested.

jrmadsen replied:
For any of the options w/ both USE and BUILD variants (python, caliper, gotcha, and ompt), if `TIMEMORY_USE_<PACKAGE>=OFF` then `TIMEMORY_BUILD_<PACKAGE>` is ignored. USE w/o BUILD means find external installation, USE + BUILD means build it from submodule.

adamjstewart replied:
With Spack, we try to never build vendored copies of software, or to link to system libraries. If a variant is on, it should link to a Spack installation. If it's off, it shouldn't link to anything at all. If the dependency is mandatory, there shouldn't even be a variant for it. This prevents packages from linking to system libraries, leading to non-deterministic builds. It also prevents packages from being built with multiple versions of the same library in the DAG. So if I understand you correctly, we never want to build from a submodule, we either want to link to an external Spack installation or not link at all.

jrmadsen replied:
So I am not trying to link to any system libraries here. "Python" in this context is actually just a template library used to generate the bindings, pybind11, and I will never let spack provide that because there is a 0% chance of it breaking any other spack build/install. 

The gotcha submodule can easily be removed, it was added bc I added the others. 

The ompt submodule can also be removed. I didn't realize there was a spack package for that which worked (`llvm-openmp-ompt`) because `ompt-openmp` is outdated. Is there a syntax for "use this config if not explicitly specified but allow another config?" E.g. if they just run `spack install timemory +ompt` then I want `depends_on("llvm-openmp-ompt+standalone")` but if they explicitly want to use `llvm-openmp-ompt~standalone` (the default), then that is fine. 

Basically the situation above is what prompted this whole submodule additions. I found it quite difficult to just do `spack install timemory +caliper` because timemory doesn't require a specific papi version but Caliper does so I would just like to defer to caliper setting the papi version. That was just the first issue, many more happened and the specs got extremely detailed, especially once TAU got involved. Basically, timemory doesn't really care about the spec of Caliper, TAU, or any similar libraries for which it provides wrappers (arm-forge, etc.). Those libraries provide generic interfaces to call their library and their corresponding timemory components just invoke those generic interfaces, e.g. `tau_marker::start()` just calls `Tau_start(...)`. Is there a way to say "if there was a previous spack install for package X, just use that installation" _without having to write a packages.yaml_? And, as mentioned previously, "use this spec as the default unless otherwise specified".

adamjstewart replied:
> Is there a way to say "if there was a previous spack install for package X, just use that installation" without having to write a packages.yaml?

This will be the default behavior of the new concretizer. An opt-in experimental version of this concretizer is expected to be released sometime in May (Spack 0.15) with a full replacement of the old concretizer scheduled for Spack 0.16.

> And, as mentioned previously, "use this spec as the default unless otherwise specified".

Not currently, the `depends_on` directive is a hard requirement, there is no way to specify preferences. We've talked about adding this, but it's harder than it sounds. For now, I would let all `depends_on` statements be as relaxed as possible.

jrmadsen replied:
Thanks. I believe all the above is resolved now

jrmadsen committed:
80 char width + added extra conflicts + fixed kokkos_modules reference

jrmadsen committed:
79 char width

jrmadsen commented:
@adamjstewart For my end, this is ready to merge

jrmadsen committed:
Removed submodule builds

- added allinea_map variant
- disabled caliper by default
- added ompt_{standalone,llvm} variants

alalazo commented:
@jrmadsen There are `flake8` errors in this PR. You can check them by following the link to Travis in the checks:
```
var/spack/repos/builtin/packages/timemory/package.py:80: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/timemory/package.py:144: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/timemory/package.py:145: [E128] continuation line under-indented for visual indent
```

jrmadsen committed:
flake8 updates

jrmadsen committed:
79 char width
